allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -209,7 +230,12 @@ def read(self, file_path: str) -> Dataset:
                 )
 
             # And finally we write to the cache if we need to.
-            if cache_file and not os.path.exists(cache_file):
+            if (
+                self.max_instances is None
```

Comment:
This line made me pause and wonder if it was correct.  I think I see why you did this (if you've set that, you're probably testing something, and don't want to cache only a part of the data as if it's the whole thing), but it's not obvious at first glance, so a comment explaining why it's there would be nice.

An alternative to having this check here would be to move the caching logic to above the place where you only keep `max_instances` (or move the slicing to below this).  I'd probably vote for that option, instead of this.  Well, that then might defeat the point of the `max_instances`, because you'd be reading all of the data...  I guess it depends on which flag you think takes precedence.  There's a fair argument that `max_instances` should take precedence, and this check should stay as it is.  In that case, noting this in the docstring (that `max_instances` disables saving data to the cache) would be good.

Reply:
I added this because I didn't want to write 500 instances to the cache, and then treat them as the whole dataset when reading the cache. I think the old code would do that.

There are two ways to avoid that. We could make the number of instances part of the filename of the cache, or we could never cache when `max_instances` is set. The latter seemed easier. It takes almost no time to read 500 instances anyways.

Reply:
Yes, I agree, I ended up at the same place at the end of my comment.  We should just add a comment and update the docstring to make this clear.

Reply:
Done